### PR TITLE
Add customizability of locale data to principles.md

### PR DESCRIPTION
**Pull request by sffc**
I made this PR to address https://github.com/unicode-org/icu4x/pull/3426#issuecomment-1545631319 since it seems we don't have the reasons behind mutable locale data written down anywhere.

### CODE REVIEW

**Comment by zbraniecki** [comment]
My assumption was that the *way* we enable this feature is by allowing custom DataProviders that overlay subset of data and route the rest to original DataProvider.

This does not require data mutability within a single DataProvider's data.

Am I misaligned?

**Comment by robertbastian** [comment]
> This does not require data mutability within a single DataProvider's data.

It does in a way. Changing the value of a string field in a data struct is easy if the struct can own the string in a `Cow<'data, str>`, and harder if it needs to borrow as `&'data str`. In the latter case the string needs to be borrowed from the `DataPayload`'s cart, which is fairly unergonomic at the moment (in fact, our cart is always `(Box<[u8]>`). We could make this more ergonomic instead of mandating that data struct fields need to be owned-or-borrowed.

**Comment by zbraniecki** [comment]
> It does in a way. Changing the value of a string field in a data struct is easy if the struct can own the string in a Cow<'data, str>, and harder if it needs to borrow as &'data str.

I was not aware of an architectural direction of allowing to modify data within a DataProvider's data.

I was exploring that in context of data within DateTimeFormat - where the OptionsBag could affect skeleton, and I'd like to store such **potentially-modified-skeleton** (say, `j` replaced with kKhH).

But for *customer* who wants a different skeleton, I envisioned they'd establish "MyDataProvider" that returns custom skeleton and routes all other requests to CLDRDataProvider. That doesn't require mutability of data, correct?

**Comment by sffc** [comment]
> But for _customer_ who wants a different skeleton, I envisioned they'd establish "MyDataProvider" that returns custom skeleton and routes all other requests to CLDRDataProvider. That doesn't require mutability of data, correct?

Right, but the customer likely will want to "clone" the default data, make some changes, and return the result. This is what I mean by "mutable"; maybe there's a better word.

**Comment by zbraniecki** [comment]
> Right, but the customer likely will want to "clone" the default data, make some changes, and return the result. This is what I mean by "mutable"; maybe there's a better word.

My assumption is that they would generate the data with some transformation in the process, and store it as postcard/baked.

You're suggesting a more dynamic model where they apply the transformation in-memory on original data, correct?

**Comment by sffc** [comment]
I added a commit to elaborate on this point of runtime-patched data.

**Comment by robertbastian** [comment]
I think the use cases for mandating this are a bit weak at the moment. I agree with Zibi that with our existing providers (databake and serde) it makes more sense to have clients use custom source data and not do any runtime tailoring. The mutability you're describing is only required with providers that clients don't control, which we don't have yet. I think for things like OS-providers, the ownership question is still completely open (currently our `DataPayload` can only borrow from a `[u8]`), and we can deal with mutability when we get there.

I also think this orthogonal to https://github.com/unicode-org/icu4x/pull/3426#issuecomment-1545631319.

**Comment by sffc** [comment]
> The mutability you're describing is only required with providers that clients don't control, which we don't have yet.

Not true:

1. The BlobProvider already supports loading data from an external source
2. The BakedProvider may be shared by many different applications using the same monorepo (see google)

**Comment by robertbastian** [comment]
> The BlobProvider already supports loading data from an external source

I would consider that client-controlled, the clients can decide which data with which tailorings to load.

> The BakedProvider may be shared by many different applications using the same monorepo (see google)

How would tailoring in such a single-provider environment look like? You always need some kind of wrapping provider, and then you can just have a second databake provider.

**Comment by sffc** [comment]
> > The BlobProvider already supports loading data from an external source
> 
> I would consider that client-controlled, the clients can decide which data with which tailorings to load.

The point is that the application ("client") may have different requirements than the data source for specific pieces of data; say they are loading the data from a CDN which only provides vanilla CLDR.

Also make note of the following use case, as currently written in the PR: "User-specific settings may override items such as datetime or decimal separators". These are not things that can be pre-computed when the application is being compiled.

> > The BakedProvider may be shared by many different applications using the same monorepo (see google)
> 
> How would tailoring in such a single-provider environment look like? You always need some kind of wrapping provider, and then you can just have a second databake provider.

It looks like this: https://github.com/unicode-org/icu4x/blob/main/docs/tutorials/data_provider.md#overwriting-specific-data-items

File: docs/design/principles.md [hunk 1]
```diff
@@ -34,6 +34,28 @@ All ICU4X code must conform to the [style guide](../process/style_guide.md), inc
 
 *Why:* One of the big problems for existing ICU users is that ICU data cannot be shared among different versions of code, forcing clients to carry hefty duplicates with small deltas.
 
+## Mutability of Locale Data
```
Comment (echeran) [inline docs/design/principles.md:37]:
```suggestion
## Customizability of Locale Data
```

**Comment by sffc** [comment]
Discussion:

- @robertbastian - I agree with the principle, but I think I don't agree on the mechanism
- @zbraniecki - It seems like there is a performance tradeoff when we add data mutability
- @sffc - We have established mechanisms for working around performance and memory issues with Cows. It is not a bottleneck.
- @robertbastian - We have other mechanisms, like Yoke
- @zbraniecki - I think it's important to exemplify the value proposition of this tradeoff: why are we forgoing some performance to mutate data at runtime?
- @Manishearth - We have not made major tradeoffs in terms of read performance in ZeroVec and VarZeroVec. In fact VarZeroVec is slow to mutate. We're very much on the "read and deserialization must be fast" side of the spectrum
- @zbraniecki - Is there a heap/stack tradeoff?
- @Manishearth - Not in very many places; but there are places in the datetime data, like Solar12 for months. It increases stack size, but it has nothing to do with mutation.
- @Manishearth - Yoke is not set up for this use case.
- @robertbastian - I don't think JSON should be a high priority
- @robertbastian - We could make the Cart be `Box<dyn Drop>`.
- @sffc - Nested `dyn Drop` has problems. vtables are bad, etc.
- @robertbastian - We could work toward making `Cow<str>` more efficient
- @sffc - I think it is already better on nightly. Not sure which change made it beter.

Conclusion:

- @sffc to address @zbraniecki's feedback in the context of this PR
- @robertbastian's comments will be decided in the context of https://github.com/unicode-org/icu4x/issues/3413

**Comment by sffc** [comment]
I think I addressed Elango and Zibi's comments. Let me know if there's anything else to do on this. I'd like an approval.

**Review by Manishearth — DISMISSED** [review]

File: docs/design/principles.md [hunk 2]
```diff
@@ -34,6 +34,30 @@ All ICU4X code must conform to the [style guide](../process/style_guide.md), inc
...
+
+Changes to data often need to happen at runtime because:
+
+1. Application-specific overrides are independent of the central data (from the operating system, for example)
+2. The data might be dynamically generated, such as application/user preferences
+
+ICU4C/ICU4J exposes certain pieces of data through user-facing APIs such as DateFormatSymbols. ICU4X departs from this approach (making the data resources modifiable when being loaded) to solve the following problems:
+
+1. User-facing APIs do not cover all data that users may need to mutate; those APIs must be perpetually maintained and updated
+2. Mutating data is a power-user feature; putting it front and center temps users to mutate it in ways they shouldn't
```
Comment (robertbastian) [inline docs/design/principles.md:56]:
```suggestion
2. Mutating data is a power-user feature; putting it front and center tempts users to mutate it in ways they shouldn't
```

**Review by robertbastian — APPROVED** [review]

**Review by robertbastian — APPROVED** [review]

**Review by zbraniecki — APPROVED** [review]

**Pull request merged**